src/extension/rpc.ts

```diff
@@ -0,0 +1,191 @@
+import { NoInfer, SafeAny } from "../types";
+import { MessageAdapter } from "./messageAdapters";
+import {
+  ApolloClientDevtoolsRPCMessage,
+  MessageType,
+  isRPCMessage,
+} from "./messages";
+
+type RPCParams = Record<string, unknown>;
+
+type RPCRequestMessageFormat = {
+  type: string;
+  params: RPCParams;
+};
+
+type RPCResponseMessageFormat =
+  | { sourceId: number; result: unknown }
+  | { sourceId: number; error: unknown };
+
+export type RPCMessage<
+  Name extends string,
+  Params extends RPCParams,
+  ReturnType,
+> = {
+  __name: Name;
+  __params: Params;
+  __returnType: ReturnType;
+};
+
+export interface RpcClient<
+  Messages extends RPCMessage<string, RPCParams, SafeAny>,
+> {
+  request: <TName extends Messages["__name"]>(
+    name: TName,
+    params: Extract<Messages, { __name: TName }>["__params"],
```

Comment:
I'm mostly happy with this, but this `params` is still a bit annoying. Right now this argumenet is required, even though its perfectly valid to send rpc calls without arguments (see the [`getClientOperations` call](https://github.com/apollographql/apollo-client-devtools/pull/1256/files#diff-4462d2ba22ebc0001138062ca5b0beef9eecbbce411838d1908bf9c4e5c2ec86R87) for example.) I'd like to make `params` only required if the params type has required fields. We will need to reorganize the function signature regardless since its awkward to set `options` with this function signature when `params` are optional.

Reply:
I ended up going with a bit different approach which avoids the awkwardness but also gives the more natural feel of allowing params to be a parameter array.

```ts
type Message = {
  add(x: number, y: number): number
}

const result = await client.withTimeout(1000).request('add', 1, 2);
```

Let me know what you think!